Expose services for ROS2 Package

lqnnguyen opened the pull request:
Expose services for Force/Torch Sensor ROS2 drive package 

Jira issue: [NXT03-333](https://samxl.atlassian.net/browse/NXT03-333)
## Motivation and Context
While the `ros1` branch of `netft_utils` provides utilities to configure the F/T sensor as services (e.g, `setBias`,`compensateGravity`), however these features are missing in `ros2` branch.  This MR addresses this issue.

## Checklist

- [x] No sensitive data is exposed in the code
- [ ] Code changes are covered by tests


## Additional Notes
Additional reviewers: @aashish-tud 


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Introduced a new utility node executable that can be launched via a command.
  - Added a new shared library for filtering functionality.
- **Documentation**
  - Updated usage instructions in the README to include the new utility node.
- **Refactor**
  - Improved logging consistency and callback binding style.
  - Updated class inheritance for clarity and maintainability.
  - Restructured the main function for better execution flow.
- **Chores**
  - Added a new package dependency to support geometry message conversions.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lqnnguyen committed:
Add logger

lqnnguyen committed:
Fix rclcpp::Node inheritance issue and include tf2_geometry_msgs

lqnnguyen committed:
Add logger as argument

lqnnguyen committed:
Add executable fot netft_utils

lqnnguyen committed:
Expose services for ros2 package

lqnnguyen committed:
Update usage for netft_utils

erichlf commented on netft_utils/CMakeLists.txt:
```diff
@@ -37,12 +39,18 @@ add_library(netft_rdt_driver SHARED
 )
 target_compile_features(netft_rdt_driver PUBLIC cxx_std_17)
 target_include_directories(netft_rdt_driver
-  PUBLIC
```
nit: I think the previous formatting was easier to read.

erichlf replied:
nit: I think the previous formatting was easier to read.

erichlf commented on netft_utils/include/lpfilter.h:
```diff
@@ -10,8 +10,9 @@ class LPFilter
   LPFilter(double deltaT, double cutoffFrequency, int numElements);
   bool update(std::vector<double> input, std::vector<double>& output);
-
+  
```
nit: stray whitespace

erichlf commented on netft_utils/src/lpfilter.cpp:
```diff
@@ -59,20 +59,20 @@ LPFilter::LPFilter(double deltaT, double cutoffFrequency, int numElements):
     out1.resize(noElements);
     out2.resize(noElements);
-    RCLCPP_INFO_STREAM("cutoffFrequency: " << cutoffFrequency << ". omega_a: " << omega_a << ". den: " << den << ". a0: " << a0 << ". a1: " << a1 << ". a2: " << a2 << ". b1: " << b1 << ". b2: " << b2);
+    // RCLCPP_INFO_STREAM("cutoffFrequency: " << cutoffFrequency << ". omega_a: " << omega_a << ". den: " << den << ". a0: " << a0 << ". a1: " << a1 << ". a2: " << a2 << ". b1: " << b1 << ". b2: " << b2);
```
nit: why not remove this? Code shouldn't be commented and left it should be deleted.

erichlf commented on netft_utils/src/netft_utils.cpp:
```diff
@@ -68,23 +68,70 @@ namespace netft_utils
 
     //Subscribe to the NetFT topic.
-    raw_data_sub = create_subscription<geometry_msgs::msg::WrenchStamped>("netft_data",100, &NetftUtils::netftCallback);
+    raw_data_sub = create_subscription<geometry_msgs::msg::WrenchStamped>(
+    "netft_data", 100,
+    [this](const geometry_msgs::msg::WrenchStamped::SharedPtr msg) { this->netftCallback(msg); });
```
```suggestion
    raw_data_sub = create_subscription<geometry_msgs::msg::WrenchStamped>(
      "netft_data", 100,
      [this](const geometry_msgs::msg::WrenchStamped::SharedPtr msg) { this->netftCallback(msg); }
    );
```

erichlf commented on netft_utils/src/netft_utils.cpp:
```diff
@@ -68,23 +68,70 @@ namespace netft_utils
     listener = new tf2_ros::TransformListener(bufferCore);
```
Has a linter been run on this?

lqnnguyen replied:
I tried to run clang-format but it just created tons of changes to the original code. That might make you overwhelmed with formating changes

erichlf replied:
If it is a separate commit then I can look at the other commits.

erichlf commented on netft_utils/src/netft_utils.cpp:
```diff
@@ -442,61 +489,63 @@ namespace netft_utils
-    return 0;
+  // Main ros loop
+  rclcpp::Rate loop_rate(500);
+  //ros::Time last;
```
nit: please remove commented code.

erichlf reviewed (changes requested):
The biggest issue is really that a linter should be run on the code.

erichlf commented:
@coderabbitai review

erichlf commented on netft_utils/src/netft_utils.cpp:
```diff
@@ -442,61 +489,63 @@ namespace netft_utils
+  std::string ft_frame;
+  double forceMaxU = 0.0;
+  double torqueMaxU = 0.0;
+  if(argc<3)
```
maybe checkout `boost::program_options`, `cxxopts`, or `getop`?

coderabbitai[bot] committed:
📝 Add docstrings to `expose_ros2_services`

Docstrings generation was requested by @erichlf.

* https://github.com/sam-xl/netft_utils/pull/1#issuecomment-2853532116

The following files were modified:

* `netft_utils/include/lpfilter.h`
* `netft_utils/include/netft_utils.h`
* `netft_utils/src/lpfilter.cpp`
* `netft_utils/src/netft_utils.cpp`

lqnnguyen committed:
Merge pull request #2 from sam-xl/coderabbitai/docstrings/etEgo15voYRmEuJ0NQO8wfncbth

📝 Add docstrings to `expose_ros2_services`

lqnnguyen committed:
Apply clang-formartting
